# Lesson 1

## Library loading

In [1]:
import pandas as pd
import numpy as np
pd.set_option('display.max_columns', None)
import warnings
warnings.filterwarnings('ignore')
import matplotlib.pyplot as plt
import seaborn as sns 

## Feature selection based on variance.

Here we will use variance threshold to remove some features that very low variance, ie those columns have almost the same values in all the rows. Hence such features do not explain anything about the target variable 

In [2]:
numerical = pd.read_csv('numerical.csv')
categorical = pd.read_csv('categorical.csv')
targets = pd.read_csv('target.csv')

In [3]:
from sklearn.feature_selection import VarianceThreshold

Remember the definition of the variance:

$\sigma_{col}^{2}=\frac{1}{n}\sum_{i=1}^{n}(x_{i}-\bar{X})^2$

In [4]:
sel = VarianceThreshold(threshold=(.9))
temp = sel.fit_transform(numerical)
temp = pd.DataFrame(temp)
print(numerical.shape)
print(temp.shape)

(95412, 315)
(95412, 305)


As it can be clearly seen, the VarianceThreshold method reduces the size of the dataframe droping those columns for which the variance is smaller than 0.9.

To check which columns were removed we can do:

In [5]:
list(sel.get_support())

[True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 False,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True

In [6]:
list(sel.variances_)

[909809.7483785483,
 208.8158971989076,
 2.8881932699556185,
 7.320489004283561,
 86.61745725511418,
 25.65924724942382,
 131.57223253672964,
 227.9397116710221,
 313.6076948242227,
 19.286818754804564,
 26.281939824591404,
 17.265241830723728,
 8.311701637927996,
 32984198.86101045,
 2126043.3424667185,
 4507490.409069261,
 2249.6547195922,
 974.8079246455828,
 1603.0129942634196,
 30.97534865272938,
 33.27364006327841,
 441.57831545183956,
 278.5789169921517,
 12.03604563865652,
 49.980973211488966,
 190.05348704784367,
 0.4614096996326538,
 4.9825142559714894,
 6.419788645235147,
 5.332371515424055,
 1.007061136251383,
 1.1111593939063544,
 1.368116911416873,
 128.44765349637763,
 1.6713672732722233,
 10.295326351943883,
 11.32502530691183,
 69.48179133914489,
 68.05635235184752,
 65.75686276277202,
 52.72468559024539,
 48.441176213956375,
 47.3710295874623,
 56.47911443932282,
 67.7692564686937,
 40.71093052337281,
 33.95626692614386,
 36.82299864542428,
 60.64881975087444,
 38.576

In [7]:
my_list_of_tuples = [(index,value) for (index,value) in enumerate(list(sel.variances_)) if value < 0.9]

In [8]:
cols_nul_variance =  [list(numerical.columns)[index] for index in [x[0] for x in my_list_of_tuples]]
cols_nul_variance

['ETH6',
 'TPE6',
 'TPE7',
 'ANC5',
 'ANC6',
 'ANC11',
 'ANC15',
 'HC15',
 'MHUC2',
 'HPHONE_D']

## Activity 1

Fill the missing code:


In [9]:
X = numerical
y = targets['TARGET_B']

In [10]:
targets.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 95412 entries, 0 to 95411
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   TARGET_B  95412 non-null  int64  
 1   TARGET_D  95412 non-null  float64
dtypes: float64(1), int64(1)
memory usage: 1.5 MB


In [11]:
from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import chi2

# Beware of the dog!!! 
kbest = SelectKBest(chi2, k=10).fit_transform(X, y)
# Here we chose 10 so that is easier to analyze results later, as we will see
selected = pd.DataFrame(kbest)
selected.head()

0       1       2       3       4       5       6        7      8  \
0  0.0   992.0   264.0   479.0   635.0     0.0   318.0  12883.0  240.0   
1  1.0  3611.0   940.0  5468.0  5218.0  4480.0  1096.0  36175.0   47.0   
2  1.0  7001.0  2040.0   497.0   546.0     0.0   292.0  11576.0  202.0   
3  0.0   640.0   160.0  1000.0  1263.0  9340.0   388.0  15130.0  109.0   
4  0.0  2520.0   627.0   576.0   594.0  5000.0   250.0   9836.0  254.0   

          9  
0   95515.0  
1  148535.0  
2   15078.0  
3  172556.0  
4    7112.0

In [12]:
# To check the scores
model = SelectKBest(chi2, k=10).fit(X, y)
df =pd.DataFrame(data = model.scores_, columns = ['score'])
df['Column'] = numerical.columns
print(df.sort_values(by = ['score'], ascending = False).head(10))
print()

cols = df.sort_values(by = ['score'], ascending = False).head(10)['Column']
print(cols)

             score    Column
311  527716.426176  CONTROLN
140  187983.976667       IC5
83    49855.611718       HV1
84    49561.067003       HV2
0     39087.069814     TCODE
133   26891.429352       MSA
13    17167.230879    POP901
137    2921.367106       IC2
14     2811.233301    POP902
303    2756.199364  RAMNTALL

311    CONTROLN
140         IC5
83          HV1
84          HV2
0         TCODE
133         MSA
13       POP901
137         IC2
14       POP902
303    RAMNTALL
Name: Column, dtype: object


In [13]:
display(selected)

0        1       2       3       4       5       6        7       8  \
0      0.0    992.0   264.0   479.0   635.0     0.0   318.0  12883.0   240.0   
1      1.0   3611.0   940.0  5468.0  5218.0  4480.0  1096.0  36175.0    47.0   
2      1.0   7001.0  2040.0   497.0   546.0     0.0   292.0  11576.0   202.0   
3      0.0    640.0   160.0  1000.0  1263.0  9340.0   388.0  15130.0   109.0   
4      0.0   2520.0   627.0   576.0   594.0  5000.0   250.0   9836.0   254.0   
...    ...      ...     ...     ...     ...     ...     ...      ...     ...   
95407  1.0  27380.0  7252.0   988.0  1025.0   380.0   481.0  18807.0    25.0   
95408  1.0   1254.0   322.0  1679.0  1723.0  3360.0   836.0  26538.0    20.0   
95409  1.0    552.0   131.0   376.0   377.0  4040.0   264.0  12178.0    58.0   
95410  0.0   1746.0   432.0  2421.0  2459.0  8735.0   544.0  15948.0   498.0   
95411  2.0   3935.0  1192.0   938.0  1049.0  9200.0   474.0  16699.0  3775.0   

              9  
0       95515.0  
1      148535.0  
2       15078.0  
3      172556.0  
4        7112.0  
...         ...  
95407  184568.0  
95408  122706.0  
95409  189641.0  
95410    4693.0  
95411  185114.0  

[95412 rows x 10 columns]

In [14]:
display(numerical)

TCODE        AGE  INCOME  WEALTH1  HIT  MALEMILI  MALEVET  VIETVETS  \
0          0  60.000000       5        9    0         0       39        34   
1          1  46.000000       6        9   16         0       15        55   
2          1  61.611649       3        1    2         0       20        29   
3          0  70.000000       1        4    2         0       23        14   
4          0  78.000000       3        2   60         1       28         9   
...      ...        ...     ...      ...  ...       ...      ...       ...   
95407      1  61.611649       5        9    0        14       36        47   
95408      1  48.000000       7        9    1         0       31        43   
95409      1  60.000000       5        9    0         0       18        46   
95410      0  58.000000       7        9    0         0       28        35   
95411      2  80.000000       5        8    3         1       33        22   

       WWIIVETS  LOCALGOV  STATEGOV  FEDGOV  WEALTH2  POP901  POP902  POP903  \
0            18        10         2       1        5     992     264     332   
1            11         6         2       1        9    3611     940     998   
2            33         6         8       1        1    7001    2040    2669   
3            31         3         0       3        0     640     160     219   
4            53        26         3       2        9    2520     627     761   
...         ...       ...       ...     ...      ...     ...     ...     ...   
95407        11         7         8      13        9   27380    7252   10037   
95408        19         4         1       0        9    1254     322     361   
95409        20         7        23       0        9     552     131     205   
95410        20         9         1       1        7    1746     432     508   
95411        27         8        10       3        8    3935    1192    1342   

       POP90C1  POP90C2  POP90C3  POP90C4  POP90C5  ETH1  ETH2  ETH3  ETH4  \
0            0       35       65       47       53    92     1     0     0   
1           99        0        0       50       50    67     0     0    31   
2            0        2       98       49       51    96     2     0     0   
3            0        8       92       54       46    61     0     0    11   
4           99        0        0       46       54     2    98     0     0   
...        ...      ...      ...      ...      ...   ...   ...   ...   ...   
95407       99        0        0       50       50    78    10     6     4   
95408       96        0        4       51       49    91     3     0     2   
95409       99        0        0       53       47    82    14     0     1   
95410       99        0        0       47       53    92     1     1     5   
95411       99        0        0       48       52    86    12     0     1   

       ETH5  ETH6  ETH7  ETH8  ETH9  ETH10  ETH11  ETH12  ETH13  ETH14  ETH15  \
0        11     0     0     0     0      0      0      0     11      0      0   
1         6     4     2     6     4     14      0      0      2      0      1   
2         2     0     0     0     0      0      0      0      2      0      0   
3        32     6     2     0     0      0      0      0     31      0      0   
4         1     0     0     0     0      0      0      0      0      0      0   
...     ...   ...   ...   ...   ...    ...    ...    ...    ...    ...    ...   
95407     5     0     0     0     1      1      0      0      3      1      0   
95408     6     1     0     1     0      0      0      0      5      0      0   
95409     9     0     0     0     0      0      0      0      9      0      0   
95410     8     0     1     2     0      1      0      0      5      0      0   
95411     1     0     0     0     0      0      0      0      0      0      0   

       ETH16  AGE901  AGE902  AGE903  AGE904  AGE905  AGE906  AGE907  CHIL1  \
0          0      39      48      51      40      50      54      25     31   
1          4      34      41      43     

# Lesson 2

## Recursive feature elimination

In [15]:
from sklearn.feature_selection import RFE
from sklearn import linear_model

# We can use any model, LM, K-NN,...
# For classification problems, the metric used to compare models is Accuracy
# For regression problems, the metric used to compare models is RMSE
#lm = linear_model.LinearRegression()
#rfe = RFE(lm, n_features_to_select=20, verbose=False)
#rfe.fit(X, y)

KeyboardInterrupt: 

In [ ]:
# After we run the algorithm, it labels the top features as 1 and the rest are marked in an increasing order of importance. 
df = pd.DataFrame(data = rfe.ranking_, columns=['Rank'])
df['Column_name'] = X.columns
df[df['Rank']==1]

## Activity 2

# Lesson 3

# Lesson 4

## Hypothesis testing

We want to test if our the **sample mean** is not equal to the **population mean** = 80.94. We also know that our **sample** has a size of 25 individuals.

$t = \frac{(\bar{X}-\mu)}{\hat{\sigma}/\sqrt{n}}$

where:

* $\bar{X}$ is the **sample mean**
* $\mu$ is the **population mean**
* $\hat{\sigma}$ is the **sample standard deviation**
* $n$ is the number of measures in our sample

In [16]:
import math

sample_mean = 80.94
pop_mean = 85
sample_std = 11.6
n = 25
statistic = (sample_mean - pop_mean)/(sample_std/math.sqrt(n))
print("Statistic is: ", statistic)

Statistic is:  -1.750000000000001


In [17]:
from scipy import stats
from numpy.random import normal


samples = {}

for i in range(10):
    sample_name = "sample_" + str(i)
    samples[sample_name] = normal(loc = 80.94, scale = 11.6, size = 25)
    sample_mean = "sample_" + str(i) + "_mean"
    samples[sample_mean] = np.mean(samples[sample_name])
    sample_std = "sample_" + str(i) + "_std"
    samples[sample_std] = np.std(samples[sample_name],ddof=1)
    sample_statistic = "sample_" + str(i) + "_t-statistic"
    samples[sample_statistic] = (samples[sample_mean]- pop_mean)/(samples[sample_std]/math.sqrt(n)) 
    print("The t-statistic for the sample {} is: {}".format(i,samples[sample_statistic]))


The t-statistic for the sample 0 is: 0.3376099975615978
The t-statistic for the sample 1 is: -0.8190004859471183
The t-statistic for the sample 2 is: -1.010376051572296
The t-statistic for the sample 3 is: -3.020069246610508
The t-statistic for the sample 4 is: -2.485459597454204
The t-statistic for the sample 5 is: -1.6049801727534176
The t-statistic for the sample 6 is: -0.5963748183060341
The t-statistic for the sample 7 is: -3.262474844084132
The t-statistic for the sample 8 is: -0.9771684709146745
The t-statistic for the sample 9 is: -2.5392166408218877


Now that we have the t-statistic for each random sample, let's make the two tails test. Why two tails? Because we are looking what is the probability that we get a **sample mean** which deviates from the **population mean** more than out t-statistic. We don't care if the our **sample mean** is bigger or smaller than the **population mean**.

Therefore, we can ask ourselves what is the probability of having a deviation within -t and t.


In [18]:
print("Assuming a significance level of 0.05")
print()

for i in range(10):
    sample_name = "sample_" + str(i)
    print("The p-value of sample {} is: {:-5.3}".format(i,stats.ttest_1samp(samples[sample_name],85)[1]))
    if ( stats.ttest_1samp(samples[sample_name],85)[1] < 0.05 ):
        print("Therefore we discard the null hypothesis Ho, as it's very unlikely to get sample {} given Ho.".format(i))
    print()

Assuming a significance level of 0.05

The p-value of sample 0 is: 0.739

The p-value of sample 1 is: 0.421

The p-value of sample 2 is: 0.322

The p-value of sample 3 is: 0.00592
Therefore we discard the null hypothesis Ho, as it's very unlikely to get sample 3 given Ho.

The p-value of sample 4 is: 0.0203
Therefore we discard the null hypothesis Ho, as it's very unlikely to get sample 4 given Ho.

The p-value of sample 5 is: 0.122

The p-value of sample 6 is: 0.557

The p-value of sample 7 is: 0.0033
Therefore we discard the null hypothesis Ho, as it's very unlikely to get sample 7 given Ho.

The p-value of sample 8 is: 0.338

The p-value of sample 9 is: 0.018
Therefore we discard the null hypothesis Ho, as it's very unlikely to get sample 9 given Ho.



# 7.04 Activity 1
    The datasets are provided in the files_for_activities folder.

    For the example covered in the lab for the last lesson, using the test statistics and then using the p-value, do you reject the null hypothesis or you fail to reject the null hypothesis?

    Here is the question again for reference: It is assumed that the mean systolic blood pressure is μ = 120 mm Hg. In the Honolulu Heart Study, a sample of n = 100 people had an average systolic blood pressure of 130.1 mm Hg with a standard deviation of 21.21 mm Hg. Is the group significantly different (with respect to systolic blood pressure!) from the regular population?

    Set up the hypothesis test.

H0 = No difference between the means
Ha = Difference between the means



In [19]:
sample_mean = 130.1
pop_mean = 120
sample_std = 21.21
n = 100
statistic = (sample_mean - pop_mean)/(sample_std/math.sqrt(n))
print("Statistic is: ", statistic)

Statistic is:  4.761904761904759


In [20]:
samples = {}

for i in range(10):
    sample_name = "sample_" + str(i)
    samples[sample_name] = normal(loc = 130.1, scale = 21.21, size = 100)
    sample_mean = "sample_" + str(i) + "_mean"
    samples[sample_mean] = np.mean(samples[sample_name])
    sample_std = "sample_" + str(i) + "_std"
    samples[sample_std] = np.std(samples[sample_name],ddof=1)
    sample_statistic = "sample_" + str(i) + "_t-statistic"
    samples[sample_statistic] = (samples[sample_mean]- pop_mean)/(samples[sample_std]/math.sqrt(n)) 
    print("The t-statistic for the sample {} is: {}".format(i,samples[sample_statistic]))

The t-statistic for the sample 0 is: 5.98462689354867
The t-statistic for the sample 1 is: 5.85252530073868
The t-statistic for the sample 2 is: 4.621378423964536
The t-statistic for the sample 3 is: 4.0456431527584735
The t-statistic for the sample 4 is: 3.9285683301612733
The t-statistic for the sample 5 is: 4.26381491926478
The t-statistic for the sample 6 is: 4.765136462820944
The t-statistic for the sample 7 is: 3.1190040623282207
The t-statistic for the sample 8 is: 4.784422357298407
The t-statistic for the sample 9 is: 4.9237139602392395


In [21]:
print("Assuming a significance level of 0.05")
print()

for i in range(10):
    sample_name = "sample_" + str(i)
    print("The p-value of sample {} is: {:-5.3}".format(i,stats.ttest_1samp(samples[sample_name],120)[1]))
    if ( stats.ttest_1samp(samples[sample_name],120)[1] < 0.05 ):
        print("Therefore we discard the null hypothesis Ho, as it's very unlikely to get sample {} given Ho.".format(i))
    print()

Assuming a significance level of 0.05

The p-value of sample 0 is: 3.48e-08
Therefore we discard the null hypothesis Ho, as it's very unlikely to get sample 0 given Ho.

The p-value of sample 1 is: 6.29e-08
Therefore we discard the null hypothesis Ho, as it's very unlikely to get sample 1 given Ho.

The p-value of sample 2 is: 1.15e-05
Therefore we discard the null hypothesis Ho, as it's very unlikely to get sample 2 given Ho.

The p-value of sample 3 is: 0.000104
Therefore we discard the null hypothesis Ho, as it's very unlikely to get sample 3 given Ho.

The p-value of sample 4 is: 0.000158
Therefore we discard the null hypothesis Ho, as it's very unlikely to get sample 4 given Ho.

The p-value of sample 5 is: 4.6e-05
Therefore we discard the null hypothesis Ho, as it's very unlikely to get sample 5 given Ho.

The p-value of sample 6 is: 6.48e-06
Therefore we discard the null hypothesis Ho, as it's very unlikely to get sample 6 given Ho.

The p-value of sample 7 is: 0.00238
Therefore

# 7.04 Activity 3

List down all the steps in conducting a hypothesis test, and see if you are able to set up the test yourself. You are free to make any assumptions in your test as you would like.

Elements needed:

    > Null hypothesis
    > Alternate hypothesis
    > Level of significance
    > Test statistic
    > P-value

Assuming that: samples are drawns from a population which is normally distributed

    Step 1: Defining the null hypothesis - This is our assumption about the population. It is defined by H0 and in this case H0: mu = X
    Step 2: Defining the Alternate hypothesis - This means what if our assumption is not true. It is defined by Ha and in this case Ha: mu != X
    Step 3: Decide a test statistics based on the information available (t-test or z-test)
    Step 4: Level of significance: This defines the rejection region / critical region. It is defined by the greek letter 'alpha'. Usually it is 0.05.

In [22]:
sample_mean1 = 105.5
sample_std1 = 20.1
n1 = 34
sample_mean2 = 90.9
sample_std2 = 12.2
n2 = 29
pooled_sample_std = math.sqrt(((n1-1)*sample_std1**2 + (n2-1)*sample_std2**2)/(n1+n2-2))
statistic = (sample_mean1-sample_mean2)/(pooled_sample_std*math.sqrt((1/n1)+(1/n2)))
print("T Statistic is: ", statistic)

T Statistic is:  3.4101131776909535


In [23]:
t = statistic

In [24]:
print(t)

3.4101131776909535


In [25]:
from scipy.stats import t

In [26]:
# Using python to find the p value and critical value
print("P value is: ", 1- t.cdf(statistic,n1+n2-2))
print("Critical Value of z is: ", t.ppf(0.025, n1+n2-2)) #alpha is 0.05

P value is:  0.0005783712704483523
Critical Value of z is:  -1.9996235841149783


# Lab | Inferential statistics
    Instructions
    It is assumed that the mean systolic blood pressure is μ = 120 mm Hg. In the Honolulu Heart Study, a sample of n = 100 people had an average systolic blood pressure of 130.1 mm Hg with a standard deviation of 21.21 mm Hg. Is the group significantly different (with respect to systolic blood pressure!) from the regular population?

    > Set up the hypothesis test.
    > Write down all the steps followed for setting up the test.
    > Calculate the test statistic by hand and also code it in Python. It should be 4.76190. We will take a look at how to make decisions based on this calculated value.
    If you finished the previous question, please go through the code for principal_component_analysis_example provided in the files_for_lab folder .

## Set up the hypothesis test
    H0 = no significant difference between the groups
    Ha = significant difference between the groups

## Write down all the steps followed for setting up the test.

Elements needed:

    > Null hypothesis
    > Alternate hypothesis
    > Level of significance
    > Test statistic
    > P-value

Assuming that: samples are drawns from a population which is normally distributed

    Step 1: Defining the null hypothesis - This is our assumption about the population. It is defined by H0 and in this case H0: mu = X
    Step 2: Defining the Alternate hypothesis - This means what if our assumption is not true. It is defined by Ha and in this case Ha: mu != X
    Step 3: Decide a test statistics based on the information available (t-test or z-test)
    Step 4: Level of significance: This defines the rejection region / critical region. It is defined by the greek letter 'alpha'. Usually it is 0.05.

## Calculate the test statistic by hand and also code it in Python. It should be 4.76190. We will take a look at how to make decisions based on this calculated value.

In [27]:
sample_mean = 130.1
pop_mean = 120
sample_std = 21.21
n = 100
statistic = (sample_mean - pop_mean)/(sample_std/math.sqrt(n))
print("Statistic is: ", statistic)

Statistic is:  4.761904761904759


In [28]:
samples = {}

for i in range(10):
    sample_name = "sample_" + str(i)
    samples[sample_name] = normal(loc = 130.1, scale = 21.21, size = 100)
    sample_mean = "sample_" + str(i) + "_mean"
    samples[sample_mean] = np.mean(samples[sample_name])
    sample_std = "sample_" + str(i) + "_std"
    samples[sample_std] = np.std(samples[sample_name],ddof=1)
    sample_statistic = "sample_" + str(i) + "_t-statistic"
    samples[sample_statistic] = (samples[sample_mean]- pop_mean)/(samples[sample_std]/math.sqrt(n)) 
    print("The t-statistic for the sample {} is: {}".format(i,samples[sample_statistic]))

The t-statistic for the sample 0 is: 3.30597656469128
The t-statistic for the sample 1 is: 5.104366405444165
The t-statistic for the sample 2 is: 5.153408939246276
The t-statistic for the sample 3 is: 3.8652457468721506
The t-statistic for the sample 4 is: 5.74694577988679
The t-statistic for the sample 5 is: 5.423931285230123
The t-statistic for the sample 6 is: 8.054129460504724
The t-statistic for the sample 7 is: 4.998613137925215
The t-statistic for the sample 8 is: 5.111089546774605
The t-statistic for the sample 9 is: 4.245643414506095


In [30]:
print("Assuming a significance level of 0.05")
print()

for i in range(10):
    sample_name = "sample_" + str(i)
    print("The p-value of sample {} is: {:-5.3}".format(i,stats.ttest_1samp(samples[sample_name],120)[1]))
    if ( stats.ttest_1samp(samples[sample_name],120)[1] < 0.05 ):
        print("Therefore we discard the null hypothesis Ho, as it's very unlikely to get sample {} given Ho.".format(i))
    print()

Assuming a significance level of 0.05

The p-value of sample 0 is: 0.00132
Therefore we discard the null hypothesis Ho, as it's very unlikely to get sample 0 given Ho.

The p-value of sample 1 is: 1.61e-06
Therefore we discard the null hypothesis Ho, as it's very unlikely to get sample 1 given Ho.

The p-value of sample 2 is: 1.31e-06
Therefore we discard the null hypothesis Ho, as it's very unlikely to get sample 2 given Ho.

The p-value of sample 3 is: 0.000198
Therefore we discard the null hypothesis Ho, as it's very unlikely to get sample 3 given Ho.

The p-value of sample 4 is: 1.01e-07
Therefore we discard the null hypothesis Ho, as it's very unlikely to get sample 4 given Ho.

The p-value of sample 5 is: 4.13e-07
Therefore we discard the null hypothesis Ho, as it's very unlikely to get sample 5 given Ho.

The p-value of sample 6 is: 1.84e-12
Therefore we discard the null hypothesis Ho, as it's very unlikely to get sample 6 given Ho.

The p-value of sample 7 is: 2.5e-06
Therefore

# Lab | Inferential statistics - T-test & P-value
    Instructions
    1 - We will have another simple example on two sample t test (pooled- when the variances are equal). But this time this is a one sided t-test
    In a packing plant, a machine packs cartons with jars. It is supposed that a new machine will pack faster on the average than the machine currently used. To test that hypothesis, the times it takes each machine to pack ten cartons are recorded. The results, in seconds, are shown in the tables in the file files_for_lab/machine.txt. Assume that there is sufficient evidence to conduct the t test, does the data provide sufficient evidence to show if one machine is better than the other

## Hypothesis Test

Do the data provide sufficient evidence to conclude that, on the average, the new machine packs faster? Perform the required hypothesis test at the 5% level of significance using the rejection region approach.

In [45]:
data = pd.read_csv('machine2.txt', sep="\t")

In [46]:
display(data)

New machine  Old machine
0         42.1         42.7
1         41.0         43.6
2         41.3         43.8
3         41.8         43.3
4         42.4         42.5
5         42.8         43.5
6         43.2         43.1
7         42.3         41.7
8         41.8         44.0
9         42.7         44.1

In [47]:
from scipy.stats import ttest_ind

In [51]:
stats.ttest_ind(data['New machine'], data['Old machine'])

Ttest_indResult(statistic=-3.3972307061176026, pvalue=0.0032111425007745158)

In [75]:
sample_mean1 = np.mean(data['New machine'])
sample_std1 = np.std(data['New machine'])
n1 = 10
sample_mean2 = np.mean(data['Old machine'])
sample_std2 = np.std(data['Old machine'])
n2 = 10

print("The mean of Sample 1:", round(sample_mean1,2))
print("The std of Sample 1:", round(sample_std1,2))
print("The mean of Sample 2:", round(sample_mean2,2))
print("The std of Sample 2:", round(sample_std2, 2))

The mean of Sample 1: 42.14
The std of Sample 1: 0.65
The mean of Sample 2: 43.23
The std of Sample 2: 0.71


In [76]:
pooled_sample_std = math.sqrt(((n1-1)*sample_std1**2 + (n2-1)*sample_std2**2)/(n1+n2-2))
statistic = (sample_mean1-sample_mean2)/(pooled_sample_std*math.sqrt((1/n1)+(1/n2)))

print("Pooled Standard Deviation is:", pooled_sample_std)
print("T Statistic is: ", statistic)

Pooled Standard Deviation is: 0.6806247130394253
T Statistic is:  -3.5809955894645813


In [71]:
# Using python to find the p value and critical value
print("P value is: ", 1- t.cdf(statistic,n1+n2-2))
print("Critical Value of z is: ", t.ppf(0.025, n1+n2-2)) #alpha is 0.05

P value is:  0.9989322065983063
Critical Value of z is:  -2.10092204024096


### Answer
    Are these independent samples? Yes, since the samples from the two machines are not related.

    Are these large samples or a normal population?

    We have  n1 < 30 and n2 < 30. We do not have large enough samples, and thus we need to check the normality assumption from both populations. Let's take a look at the normality plots for this data:

    From the normal probability plots, we conclude that both populations may come from normal distributions. Remember the plots do not indicate that they DO come from a normal distribution. It only shows if there are clear violations. We should proceed with caution.

    Do the populations have equal variance? No information allows us to assume they are equal. We can use our rule of thumb to see if they are “close.” They are not that different as s1/s2 = 0.683/0.750  is quite close to 1. This assumption does not seem to be violated.

    We can thus proceed with the pooled t-test.

    Let  denote the mean for the new machine and  denote the mean for the old machine.

    The null hypothesis is that there is no difference in the two population means, i.e.

    The alternative is that the new machine is faster, i.e.

    The significance level is 5%. Since we may assume the population variances are equal, we first have to calculate the pooled standard deviation:

    The test statistic is:

    The alternative is left-tailed so the critical value is the value  such that , with  degrees of freedom. The critical value is -1.7341. The rejection region is .

    Our test statistic, -3.3978, is in our rejection region, therefore, we reject the null hypothesis. With a significance level of 5%, we reject the null hypothesis and conclude there is enough evidence to suggest that the new machine is faster than the old machine.

![title](https://online.stat.psu.edu/stat500/sites/stat500/files/inline-images/500%20l7%20ex%20pack%20machines%20normality%20m1.png)

![title](https://online.stat.psu.edu/stat500/sites/stat500/files/inline-images/500%20l7%20compare%20pack%20normality%20plot%20of%20old%20mach.png)

https://online.stat.psu.edu/stat500/lesson/7/7.3/7.3.1/7.3.1.1

### Exercise 2

    2 - An additional problem (not mandatory): In this case we can't assume that the population variances are equal. Hence in this case we cannot pool the variances. Independent random samples of 17 sophomores and 13 juniors attending a large university yield the following data on grade point averages. Data is provided in the file student_gpa.txt. At the 5% significance level, do the data provide sufficient evidence to conclude that the mean GPAs of sophomores and juniors at the university differ?

    Test statistics can be calculated as: link to the image - Test statistics calculation for Unpooled Variance Case

    Degrees of freedom is (n1-1)+(n2-1).

Independent random samples of 17 sophomores and 13 juniors attending a large university yield the following data on grade point averages (student_gpa.txt): [1]

    Sophomores
    3.04	2.92	2.86	1.71	3.60
    3.49	3.30	2.28	3.11	2.88
    2.82	2.13	2.11	3.03	3.27
    2.60	3.13	
    
    Juniors
    2.56	3.47	2.65	2.77	3.26
    3.00	2.70	3.20	3.39	3.00
    3.19	2.58	2.98	 	 

In [80]:
data2 = pd.read_csv('student_gpa2.txt', sep="\t")

In [81]:
display(data2)

Sophomores  Juniors
0         3.04     2.56
1         1.71     2.77
2         3.30     2.70
3         2.88     3.00
4         2.11     2.98
5         2.60     3.47
6         2.92     3.26
7         3.60     3.20
8         2.28     3.19
9         2.82     2.65
10        3.03     3.00
11        3.13     3.39
12        2.86     2.58
13        3.49      NaN
14        3.11      NaN
15        2.13      NaN
16        3.27      NaN

In [83]:
data2['Sophomores']

0     3.04
1     1.71
2     3.30
3     2.88
4     2.11
5     2.60
6     2.92
7     3.60
8     2.28
9     2.82
10    3.03
11    3.13
12    2.86
13    3.49
14    3.11
15    2.13
16    3.27
Name: Sophomores, dtype: float64

In [86]:
data2['Juniors'][:13]

0     2.56
1     2.77
2     2.70
3     3.00
4     2.98
5     3.47
6     3.26
7     3.20
8     3.19
9     2.65
10    3.00
11    3.39
12    2.58
Name: Juniors, dtype: float64

In [87]:
stats.ttest_ind(data2['Sophomores'], data2['Juniors'][:13])

Ttest_indResult(statistic=-0.864325455323425, pvalue=0.39475359666695975)

In [88]:
sample_mean3 = np.mean(data2['Sophomores'])
sample_std3 = np.std(data2['Sophomores'])
n3 = 17
sample_mean4 = np.mean(data2['Juniors'][:13])
sample_std4 = np.std(data2['Juniors'][:13])
n4 = 13

print("The mean of Sophomores:", round(sample_mean3,2))
print("The std of Sophomores:", round(sample_std3,2))
print("The mean of Juniors:", round(sample_mean4,2))
print("The std of Juniors:", round(sample_std4, 2))

The mean of Sophomores: 2.84
The std of Sophomores: 0.5
The mean of Juniors: 2.98
The std of Juniors: 0.3


In [89]:
pooled_sample_std = math.sqrt(((n3-1)*sample_std3**2 + (n4-1)*sample_std4**2)/(n3+n4-2))
statistic = (sample_mean3-sample_mean4)/(pooled_sample_std*math.sqrt((1/n3)+(1/n4)))

print("Pooled Standard Deviation is:", pooled_sample_std)
print("T Statistic is: ", statistic)

Pooled Standard Deviation is: 0.42798025956891655
T Statistic is:  -0.892728724944096


In [71]:
# Using python to find the p value and critical value
print("P value is: ", 1- t.cdf(statistic,n1+n2-2))
print("Critical Value of z is: ", t.ppf(0.025, n1+n2-2)) #alpha is 0.05

P value is:  0.9989322065983063
Critical Value of z is:  -2.10092204024096


In [ ]:
stats.ttest_ind(data['New machine'], data['Old machine'])

sample_mean1 = np.mean(data['New machine'])
sample_std1 = np.std(data['New machine'])
n1 = 10
sample_mean2 = np.mean(data['Old machine'])
sample_std2 = np.std(data['Old machine'])
n2 = 10

print("The mean of Sample 1:", round(sample_mean1,2))
print("The std of Sample 1:", round(sample_std1,2))
print("The mean of Sample 2:", round(sample_mean2,2))
print("The std of Sample 2:", round(sample_std2, 2))

pooled_sample_std = math.sqrt(((n1-1)*sample_std1**2 + (n2-1)*sample_std2**2)/(n1+n2-2))
statistic = (sample_mean1-sample_mean2)/(pooled_sample_std*math.sqrt((1/n1)+(1/n2)))

print("Pooled Standard Deviation is:", pooled_sample_std)
print("T Statistic is: ", statistic)

# Using python to find the p value and critical value
print("P value is: ", 1- t.cdf(statistic,n1+n2-2))
print("Critical Value of z is: ", t.ppf(0.025, n1+n2-2)) #alpha is 0.05

### Answer
    The response variable is GPA and is quantitative. The explanatory variable is class standing (sophomores or juniors) is categorical. The two populations are independent. Since we don't have large samples from both populations, we need to check the normal probability plots of the two samples:
    Normal Probability Plot of Sophomores
    Normality plot of the grade point averages of the sophomores.

    Normal Probability Plot of Juniors
    Normality plot of grade point averages of the juniors.

    There is no indication that there is a violation of the normal assumption for both samples. As before, we should proceed with caution.

    Now, we need to determine whether to use the pooled t-test or the non-pooled (separate variances) t-test. The summary statistics are:

    The standard deviations are 0.520 and 0.3093 respectively; both the sample sizes are small, and the standard deviations are quite different from each other. We, therefore, decide to use an unpooled t-test.

    The null and alternative hypotheses are:

    H0: μ1−μ2=0 vs Ha μ1−μ2 ≠ 0
    The significance level is 5%. Perform the 2-sample t-test in Minitab with the appropriate alternative hypothesis.

    Remember, the default for the 2-sample t-test in Minitab is the non-pooled one. Minitab generates the following output.

    Two sample T for sophomores vs juniors
        N	Mean	StDev	SE Mean
    sophomore	17	2.840	0.52	0.13
    junior	13	2.981	0.309	0.086
    95% CI for mu sophomore - mu juniors: (-0.45, 0.173)

    T-Test mu sophomore = mu juniors (Vs no =): T = -0.92

    P = 0.36 DF = 26

    Since the p-value of 0.36 is larger than α = 0.05, we fail to reject the null hypothesis.

    At 5% level of significance, the data does not provide sufficient evidence that the mean GPAs of sophomores and juniors at the university are different.

    Find a 95% confidence interval for the difference between the mean GPA of Sophomores and the mean GPA of Juniors using Minitab.
    Answer
    Minitab output:
    95% CI for mu sophomore- mu juniors is;
    (-0.45, 0.173)

    Interpreting the result:
    We are 95% confident that the difference between the mean GPA of sophomores and juniors is between -0.45 and 0.173.

    Note! When entering values into the 'Samples in different columns' input boxes, Minitab always subtracts the second value (column entered second) from the first value (column entered first).

![title](https://online.stat.psu.edu/stat500/sites/stat500/files/inline-images/500%20l7%20ex%20normality%20plot%20of%20sophomores.png)

![title](https://online.stat.psu.edu/stat500/sites/stat500/files/inline-images/500%20l7%20normality%20plot%20of%20juniors%20gpa.png)

https://online.stat.psu.edu/stat500/book/export/html/581